blue_ai/envs/transient_goals.py

```diff
@@ -52,88 +57,114 @@ def __init__(
-            self.agent_dir = (self.agent_dir + 1) % 4
-
-        # Move forward
-        elif action == self.actions.forward:
-            if fwd_cell is None or fwd_cell.can_overlap():
-                self.agent_pos = tuple(fwd_pos)
-                # reward -= 0.001
-            if fwd_cell is not None and fwd_cell.type == "goal":
-                terminated = True
-                reward = self._reward()
-                mandatory += 1
-            if fwd_cell is not None and fwd_cell.type == "goalNoTerminate":
-                reward = fwd_cell.reward
-                optional += 1
-                self.grid.set(fwd_pos[0], fwd_pos[1], None)
-
-                if self.replace_transient_goals:
-                    self.obstacles.append(GoalNoTerminate(reward=self.transient_reward))
-                    self.place_obj(self.obstacles[-1], max_tries=100)
-            if fwd_cell is not None and fwd_cell.type == "obstacleNoTerminate":
-                reward = fwd_cell.reward
-                self.grid.set(fwd_pos[0], fwd_pos[1], None)
-
-                if self.replace_transient_obstacles:
-                    self.penalties.append(
-                        ObstacleNoTerminate(reward=self.transient_penalty)
-                    )
-                    self.place_obj(self.penalties[-1], max_tries=100)
-
-            if fwd_cell is not None and fwd_cell.type == "lava":
-                terminated = True
-
-        # Done action
-        elif action == self.actions.done:
-            terminated = True
+        terminated, reward = False, 0
 
-        else:
-            raise ValueError(f"Unknown action: {action}")
+        if fwd_cell is None or fwd_cell.can_overlap():
+            self.agent_pos = tuple(fwd_pos)
+
+        if fwd_cell is None:
+            return terminated, reward
+
+        terminated = self._check_termination_conditions(fwd_cell.type)
+        reward = self._get_reward_amount(fwd_cell)
+
+        is_goal_no_terminate = fwd_cell.type == "goalNoTerminate"
+        is_obstacle_no_terminate = fwd_cell.type == "obstacleNoTerminate"
+
+        # Create a new obstacle/reward if the current one is "used"
+
+        if self.replace_transient_goals and is_goal_no_terminate:
+            self.obstacles.append(GoalNoTerminate(self.transient_reward))
+            self.place_obj(self.obstacles[-1], max_tries=100)
+
+        if self.replace_transient_obstacles and is_obstacle_no_terminate:
+            self.penalties.append(ObstacleNoTerminate(self.transient_penalty))
+            self.place_obj(self.penalties[-1], max_tries=100)
+
+        # Remove game object once traversed
+        if is_obstacle_no_terminate or is_goal_no_terminate:
+            self.grid.set(fwd_pos[0], fwd_pos[1], None)
+
+        return terminated, reward
 
-        if self.step_count >= self.max_steps:
-            truncated = True
+    def step(self, action):
+        self.step_count += 1
+
+        reward = 0
+        terminated = False
+
+        match action:
```

Comment:
This is some new syntax and bumps our minimum python version up to 3.10 

See [the release notes](https://docs.python.org/3/whatsnew/3.10.html#pep-634-structural-pattern-matching) for python 3.10.